OC-4074 Update User r1

hosh opened the pull request:
Update User chef_db (rebase-1)
- Updates admin flag
- Updates new public key
- Updates password information
- Postgresql / Mysql

Supersedes Pull Request #16

hosh committed:
[OC-4074] Added initial Postgres support for update users

hosh committed:
[OC-4074] Added update_fun() for user

hosh committed:
[OC-4074] Update only admin flag for now

hosh committed:
[OC-4074] chef_sql:update_user should persist changes to public key

hosh committed:
[OC-4074] update_user() should update password data

hosh committed:
[OC-4074] Added MySQL support

seth commented on itest/itest.erl:
```diff
@@ -1176,11 +1178,27 @@ delete_user_data() ->
+  User = make_user(<<"user07">>),
+  Username = User#chef_user.username,
+  ?assertEqual({ok, 1}, chef_sql:create_user(User)),
+
```
It might help the clarity of the test to also assert that before you make the change that admin is indeed `false`.

hosh committed:
[OC-4074] Clarified itest for update user

seth commented:
Here are the fixes to apply to get dialyzer passing. You'll also need the patch that I put on top of your PR for chef_objects that can be found in chef_objects sf/OC-4074

```
diff --git a/src/chef_db.erl b/src/chef_db.erl
index e79711b..6618120 100644
--- a/src/chef_db.erl
+++ b/src/chef_db.erl
@@ -153,7 +153,8 @@
                       'update_environment' |
                       'update_client' |
                       'update_node' |
-                      'update_role'.
+                      'update_role' |
+                      'update_user'.

 -type delete_fun() :: 'delete_cookbook_version' |
                       'delete_data_bag' |
@@ -218,7 +219,7 @@ count_user_admins(#context{reqid = ReqId}) ->
   end.

 %%%
--spec user_record_to_authz_id(any(), any()) -> id().
+-spec user_record_to_authz_id(#context{}, #chef_user{} | not_found) -> id().
 user_record_to_authz_id(#context{}, #chef_user{} = UserRecord) ->
     UserRecord#chef_user.authz_id;
 user_record_to_authz_id(#context{}, not_found) ->
@@ -728,7 +729,7 @@ update_data_bag_item(#context{}=Ctx, DataBagItem, ActorId) ->
     update_object(Ctx, ActorId, update_data_bag_item, DataBagItem).

 %% Update data for a User.
-update_user(#context{}=Ctx, User, ActorId) ->
+update_user(#context{}=Ctx, #chef_user{}=User, ActorId) ->
     update_object(Ctx, ActorId, update_user, User).

 -spec update_cookbook_version(DbContext :: #context{},
@@ -1211,10 +1212,13 @@ delete_object(#context{reqid = ReqId}, Fun, Id) ->
         Result -> Result
     end.

--spec update_object(#context{}, object_id(), update_fun(), chef_object() | #chef_cookbook_version{}) -> ok |
-                                                                             not_found |
-                                                                             {conflict, any()} |
-                                                                             {error, any()}.
+-spec update_object(#context{}, object_id(), update_fun(),
+                    chef_object() |
+                    #chef_user{} |
+                    #chef_cookbook_version{}) -> ok |
+                                                 not_found |
+                                                 {conflict, any()} |
+                                                 {error, any()}.
 %% @doc Generic update for Chef object types. `Fun' is the appropriate function in the
 %% `chef_sql' module. `Object' is a Chef object (record) with updated data.
 update_object(#context{reqid = ReqId}, ActorId, Fun, Object) ->
@@ -1227,11 +1231,6 @@ update_object(#context{reqid = ReqId}, ActorId, Fun, Object) ->
         {error, Error} -> {error, Error}
     end.

-as_bin(S) when is_list(S) ->
-    list_to_binary(S);
-as_bin(B) when is_binary(B) ->
-    B.
-
 -spec get_id(chef_object() | #chef_user{} | #chef_client{} | #chef_sandbox{} | #chef_cookbook_version{}) -> object_id().
 %% @doc Return the `id' field from a `chef_object()' record type.
 get_id(#chef_client{id = Id}) ->

```

hosh committed:
[OC-4074] Fixed specs for update_user()

 - Applied patch from @seth
   https://github.com/opscode/chef_db/pull/17#issuecomment-9498742
 - Added spec for chef_sql:update_user()

hosh committed:
[OC-4074] Cleaned up unused variable

hosh committed:
[OC-4074] Cleaned up some formatting

seth commented:
Looks good. Local tests and itests pass for me. Merge when ready.

hosh committed:
[OC-4075] Fixed spec for update_fun()

seth commented:
yep, merge :)
